### PR TITLE
[THE QUIRKENING] Adds customization options to Organ Prosthetic, Junkie, Smoker, and Alcoholic!

### PR DESCRIPTION
**Pull request by DATA-xPUNGED**
## About The Pull Request

Lets people pick:
- The shitty organ they spawn with (still has the restrictions of not being possible to get heart when you are a plasmamen, for example)
- Their favorite alcohol to blackout with
- Their favorite brand of cigs
- Their favorite brand of drugs!
A lot of the code was taken from a downstream, made by @Floofies, i already had my own implementation but it was buggy and not nearly as clean, then i was made aware of theirs.
## Why It's Good For The Game

Character Customization is good and fun and lets people have more choice in their roleplay.
## Changelog
:cl:
add: Prosthetic organ users are now allowed to pick what they want to start with.
add: Junkie of all kinds, including smokers and alcoholics, are also able to pick their favorite type.
/:cl:

### CODE REVIEW

File: code/datums/quirks/negative_quirks/junkie.dm [hunk 1]
```diff
@@ -19,11 +18,21 @@
...
 
+/datum/quirk_constant_data/junkie
+	associated_typepath = /datum/quirk/item_quirk/junkie
+	customization_options = list(/datum/preference/choiced/junkie)
+
 /datum/quirk/item_quirk/junkie/add_unique(client/client_source)
 	var/mob/living/carbon/human/human_holder = quirk_holder
 
+	// Avoid setting reagent_type for the Smoker and Alcoholic subtypes.
+	if(type == /datum/quirk/item_quirk/junkie)
```
Comment (Iamgoofball) [inline code/datums/quirks/negative_quirks/junkie.dm:29]: This should be refactored into an Addict path with subtypes for drugs, smokes, and booze

**Comment by ophaq** [comment]
Could you possibly add an option to start out with no legs with the paraplegic quirk? I have seen other code bases have this option and I think this falls in line with these changes. I would have done this already if I knew how...

**Comment by DATA-xPUNGED** [comment]
> First off, this is great because I was thinking that we really needed these changes.
> 
> Second, could you possibly add an option to start out with no legs with the paraplegic quirk? I have seen other code bases have this option and I think this falls in line with these changes. I would have done this already if I knew how...

THat is a great idea, i'll see if i muster up some courage to do it.

**Comment by ophaq** [comment]
> > First off, this is great because I was thinking that we really needed these changes.
> > Second, could you possibly add an option to start out with no legs with the paraplegic quirk? I have seen other code bases have this option and I think this falls in line with these changes. I would have done this already if I knew how...
> 
> THat is a great idea, i'll see if i muster up some courage to do it.

I would be really grateful if you could.

**Comment by Jezzared** [comment]
I'd also like to note that this is a Bounty from me! I am sponsoring this! In that sense i'd like to be involved in any discussions regarding potential balance/whatever implications. (Therefore i am commenting to subscribe to the thread and also announce this.)

DATA is super cool and has been working on this for a while. I'm very excited to see it in!

**Comment by honkpocket** [comment]
You're already pretty far into this PR so I can understand if this suggestion is disregarded, but I would personally enjoy it if the quirk which makes half of your body limp offered the choice of halves rather than being a random half every round.

**Comment by SomeoneYouProbablyKnow** [comment]
> You're already pretty far into this PR so I can understand if this suggestion is disregarded, but I would personally enjoy it if the quirk which makes half of your body limp offered the choice of halves rather than being a random half every round.

I think there is a PR just for that. #81992

**Comment by DATA-xPUNGED** [comment]
> You're already pretty far into this PR so I can understand if this suggestion is disregarded, but I would personally enjoy it if the quirk which makes half of your body limp offered the choice of halves rather than being a random half every round.

I was actually working on that one, but that other PR beat me to it

**Comment by DaCoolBoss** [comment]
If it's not out of scope to add stuff, could hooch be on the alcoholic list?

**Comment by Jezzared** [comment]
We're avoiding adding mixed drinks to the list because that rapidly becomes unsustainable.

**Comment by 00-Steven** [comment]
baller

**Comment by Profakos** [comment]
Great work! One request though, could you try somehow get rid of `GLOB.alcohol_containers`? Even making an associated list of bottles and reagent types would be better than instantiating every single alcohol bottles floating around in nullspace.

File: code/_globalvars/lists/quirks.dm [hunk 2]
```diff
@@ -10,22 +10,65 @@ GLOBAL_LIST_INIT(nearsighted_glasses, list(
...
 ))
+
+///Options for the Junkie quirk to choose from
+GLOBAL_LIST_INIT(possible_junkie_addictions, setup_junkie_addictions(list(
+		/datum/reagent/drug/blastoff,
+		/datum/reagent/drug/krokodil,
+		/datum/reagent/medicine/morphine,
+		/datum/reagent/drug/happiness,
+		/datum/reagent/drug/methamphetamine,
+	)))
```
Comment (Jacquerel) [inline code/_globalvars/lists/quirks.dm:41]:
```suggestion
GLOBAL_LIST_INIT(possible_junkie_addictions, setup_junkie_addictions(list(
	/datum/reagent/drug/blastoff,
	/datum/reagent/drug/krokodil,
	/datum/reagent/medicine/morphine,
	/datum/reagent/drug/happiness,
	/datum/reagent/drug/methamphetamine,
)))
```
Same for the others.

File: code/datums/quirks/negative_quirks/junkie.dm (hunk 1 above)
Comment (Jacquerel) [reply code/datums/quirks/negative_quirks/junkie.dm:29]: Do this please. it's silly to use inheritance this way.

File: code/datums/quirks/negative_quirks/prosthetic_organ.dm [hunk 3]
```diff
@@ -47,14 +58,14 @@
 			prosthetic = new /obj/item/organ/internal/stomach/cybernetic/surplus
 			slot_string = "stomach"
 	medical_record_text = "During physical examination, patient was found to have a low-budget prosthetic [slot_string]. \
-	<b>Removal of these organs is known to be dangerous to the patient as well as the practitioner.</b>"
+	Removal of these organs is known to be dangerous to the patient as well as the practitioner."
```
Comment (Jacquerel) [inline code/datums/quirks/negative_quirks/prosthetic_organ.dm:61]:
```suggestion
		Removal of these organs is known to be dangerous to the patient as well as the practitioner."
```

**Review by Jacquerel — CHANGES_REQUESTED** [review]
> Great work! One request though, could you try somehow get rid of `GLOB.alcohol_containers`? Even making an associated list of bottles and reagent types would be better than instantiating every single alcohol bottles floating around in nullspace.

If you could manage it in a sane way, I would also very much appreciate this.

File: code/datums/quirks/negative_quirks/addict.dm [hunk 4]
```diff
@@ -17,17 +11,19 @@
 	var/obj/item/accessory_type //! If this is null, an accessory won't be spawned.
 	var/process_interval = 30 SECONDS //! how frequently the quirk processes
 	var/next_process = 0 //! ticker for processing
-	var/drug_flavour_text = "Better hope you don't run out..."
+	var/drug_flavour_text = "Better hope you don't run out... of what, exactly? You don't know."
+	abstract_parent_type = /datum/quirk/item_quirk/addict
```
Comment (Jacquerel) [inline code/datums/quirks/negative_quirks/addict.dm:15]: Please put modifications to existing vars above var declarations.

File: code/datums/quirks/negative_quirks/addict.dm [hunk 5]
```diff
@@ -17,17 +11,19 @@
...
 
-/datum/quirk/item_quirk/junkie/add_unique(client/client_source)
+/datum/quirk/item_quirk/addict/add_unique(client/client_source)
 	var/mob/living/carbon/human/human_holder = quirk_holder
+	var/addiction
```
Comment (Jacquerel) [inline code/datums/quirks/negative_quirks/addict.dm:19]: What is this declaration for?

File: code/datums/quirks/negative_quirks/addict.dm [hunk 6]
```diff
@@ -65,12 +61,35 @@
...
+	quirk_flags = QUIRK_HUMAN_ONLY|QUIRK_PROCESSES
+	mail_goodies = list(/obj/effect/spawner/random/contraband/narcotics)
+	drug_flavour_text = "Better hope you don't run out..."
+
+/datum/quirk_constant_data/junkie
+	associated_typepath = /datum/quirk/item_quirk/addict/junkie
+	customization_options = list(/datum/preference/choiced/junkie)
+
+/datum/quirk/item_quirk/addict/junkie/add_unique(client/client_source)
+
```
Comment (Jacquerel) [inline code/datums/quirks/negative_quirks/addict.dm:81]:
```suggestion
```

File: code/datums/quirks/negative_quirks/addict.dm [hunk 7]
```diff
@@ -65,12 +61,35 @@
...
+		reagent_type = GLOB.possible_junkie_addictions[addiction]
+	return ..()
+
+/datum/quirk/item_quirk/addict/remove()
 	if(quirk_holder && reagent_instance)
 		for(var/addiction_type in subtypesof(/datum/addiction))
 			quirk_holder.mind.remove_addiction_points(addiction_type, MAX_ADDICTION_POINTS)
 
-/datum/quirk/item_quirk/junkie/process(seconds_per_tick)
+/datum/quirk/item_quirk/addict/process(seconds_per_tick)
```
Comment (Jacquerel) [inline code/datums/quirks/negative_quirks/addict.dm:92]:
```suggestion
/datum/quirk/item_quirk/addict/junkie/process(seconds_per_tick)
```
Or if it's meant to be for the abstract quirk, move it above the declaration for `junkie`

**Review by Jacquerel — CHANGES_REQUESTED** [review]
Handful of small things

File: code/datums/quirks/negative_quirks/addict.dm (hunk 5 above)
Comment (DATA-xPUNGED) [reply code/datums/quirks/negative_quirks/addict.dm:19]: I was trying to fix the TGUI i think, removing it

File: code/datums/quirks/negative_quirks/addict.dm (hunk 4 above)
Comment (DATA-xPUNGED) [reply code/datums/quirks/negative_quirks/addict.dm:15]: I'm not sure what you mean by this?

File: code/datums/quirks/negative_quirks/addict.dm (hunk 7 above)
Comment (DATA-xPUNGED) [reply code/datums/quirks/negative_quirks/addict.dm:92]: done

File: code/datums/quirks/negative_quirks/addict.dm (hunk 4 above)
Comment (Jacquerel) [reply code/datums/quirks/negative_quirks/addict.dm:15]: Things prefixed by `var/` are declarations of new vars that you are creating on this subtype.
Things not prefixed by `var` are not.
Please put all new things below all old things.

**Comment by Jezzared** [comment]
Apparently this wasn't working fully for Alcoholic yet.

**Comment by ophaq** [comment]
Just curious since this is merged now, would being able to choose being legless or not for paraplegia have to be in a new PR right? Or did you manage to find a way to do that in this?

**Comment by DATA-xPUNGED** [comment]
> Just curious since this is merged now, would being able to choose being legless or not for paraplegia have to be in a new PR right? Or did you manage to find a way to do that in this?

That was always going to be another PR, this one should have already been atomized anyways

**Comment by DATA-xPUNGED** [comment]
> Apparently this wasn't working fully for Alcoholic yet.

YEP, THE BUG IS ON LIVE NOW; Trying to set options for alcoholic will blue screen!

**Pull request merged**